pyvista/core/utilities/features.py

```diff
@@ -91,14 +96,167 @@ def voxelize(mesh, density=None, check_surface=True):
     ugrid = pyvista.UnstructuredGrid(grid)
 
     # get part of the mesh within the mesh's bounding surface.
-    selection = ugrid.select_enclosed_points(surface, tolerance=0.0, check_surface=check_surface)
+    selection = grid.select_enclosed_points(surface, tolerance=0.0, check_surface=check_surface)
     mask = selection.point_data['SelectedPoints'].view(np.bool_)
 
-    # extract cells from point indices
-    vox = ugrid.extract_points(mask)
+    vox = grid.extract_points(mask, progress_bar=progress_bar)
+
     return vox
 
 
+def voxelize_volume(mesh, density=None, check_surface=True):
+    """Voxelize mesh to create a StructuredGrid or a RectilinearGrid voxel volume.
+
+    Creates a voxel volume that encloses the input mesh and discretises the cells
+    within the volume that intersect and are contained within the input mesh. A
+    StructuredGrid or RectilinearGrid is created based on density value(s) chosen.
+
+    1. Create a StructuredGrid if ``density`` values have equal dimensions : StructuredGrid.
```

Comment:
If we aren't going to use the 'blanking' functionality of `StructuredGrid`, we should use `ImageData` when density is single value for each or return `RectilinearGrid` in both cases.  I would lean towards the latter to keep a consistent return type, but I think either choice is fine.

Reply:
I take your point of having a consistent return type, however, I'm not sure a `RectilinearGrid` is sufficient for a voxel volume that's created with an **equal** `density` because a `RectilinearGrid` is defined as a structure with variable spacing in all 3 coordinate directions. I also think if we keep the return type as a `StructuredGrid` for cases where the `density` passed are equal, this would ensure that all future implementations of `vtkStructuredGrid` functionality in `PyVista` would be available to the returned data from this function (without having to `cast_to_structured_grid()`). 

For example, implementation of the "blanking" functionality of a `StructuredGrid` is a good idea, and should be a feature. I think other [CellGhostTypes](https://vtk.org/doc/nightly/html/classvtkDataSetAttributes.html#a9881ee5ce6e9cdb8ff94ce98b4815af4) methods of `vtkStructuredGrid` would be great to include in an additional PR like `EXTERIORCELL` and `HIDDENCELL`. These would be a great addition and help the function submitted in this PR.

Reply:
I think I have confused things a lot.  Sorry.  Let me start over.

`StructuredGrid` can be used for any arbitrary structured mesh including one with a cubical shape with equal number of cells in each direction, one with nonequal number of cells, one with nonuniform cell sizes in each direction, and also a mesh that is curved.  For example, a mesh that is curved along an axis as long as the mesh connectivity is structured.  The points can be in arbitrary positions, but the cell connectivity has a structured i-j-k relationship.

`ImageData` describes a mesh that is a rectangle and also that has a single cell length along a direction, but it can have different cell lengths in each direction.  For example x-length is a constant 1, y-length is a constant 2, z-length is a constant 3.

`RectilinearGrid` is like `ImageData`, but can have nonuniform lengths in each direction.  For example, x points at `(0, 1, 5, 10)` and y points at `(1, 2, 3, 6)`, etc.  Note that this can hold the same data as `ImageData`, but it just has constantly spaced x, y , z points.

`StructuredGrid` is not really needed for this function unless we want to take advantage of the 'blanking' functionality.  If we want to go this route in the future, we should set the groundwork here to return `StructuredGrid`, so it isn't a breaking change in the future.  Or at least think it through.

If we don't want `StructuredGrid`, we can use `ImageData` as of now since we are using a constant mesh size in each direction.  If we allow nonuniform (in a 1D sense) meshes in the future, we need to use `RectilinearGrid`.

Reply:
No worries, I follow what you're saying now. 

I think having cells with a structured i-j-k relationship and connectivity would be beneficial. A `StructuredGrid` would be good to return here for this reason because the output from this could be used in further processing (e.g., fluid modelling) where other packages require the i-j-k connectivity. I also agree that this would allow future  `vtkStructuredGrid` functionality (e.g., "blanking" or "exteriorcell" etc) to be added without this function breaking. I am not a VTK expert, so I may not be best placed to add these. 

Returning a `StructuredGrid` for all density values, uniform or nonuniform, is cleaner code too.

Reply:
I think this is getting too complicated if you aren't going to implement the blanking functionality.  And to be clear, all three options give i-j-k cellular conncetivity.  `StructuredGrid` allows arbitrary hexahedron shapes for its cells, i.e. arbitrary point positions, while the other two require rectangular hexahedron shapes.   I'd recommend just returning `RectilinearGrid` in all cases.  A future PR could consider how we allow for nonuniform cell sizes in each direction for both `voxelize` and `voxelize_volume` to fully take advantage of this.

Reply:
My apologies if I have complicated this. My understanding of grid types was not as thorough as it should be; thanks for the explanations and clarifications @MatthewFlamm. I'm happy to set the return type as a `RectilinearGrid` for `voxelize_volume()` in this PR. 

I agree, it would be good to explore nonuniform cell sizes in a future PR. Voxels have great utility that we should take full advantage of.

Reply:
@MatthewFlamm @bsburnham Could we resolve this discussion and merge?